src/hotspot/share/cds/filemap.cpp

```diff
@@ -1738,19 +1738,39 @@ void FileMapInfo::close() {
     _file_open = false;
     _fd = -1;
   }
+
+  // Workers are no longer needed.
+  _archive_workers.shutdown();
 }
 
+class ArchiveRegionPretouchTask : public ArchiveWorkerTask {
+private:
+  char* const _from;
+  size_t const _bytes;
+
+public:
+  ArchiveRegionPretouchTask(char* from, size_t bytes) :
+    ArchiveWorkerTask("Archive Regions Pretouch"), _from(from), _bytes(bytes) {}
+
+  void work(int chunk, int max_chunks) override {
+    char* start = _from + MIN2(_bytes, _bytes * chunk / max_chunks);
+    char* end   = _from + MIN2(_bytes, _bytes * (chunk + 1) / max_chunks);
+    os::pretouch_memory(start, end);
```

Comment:
What happens if I have many cores and a small memory range? We would have many workers for a potentially smallish total range. Would start-end then not be tiny? On Linux, we would do madvise MADV_POPULATE_WRITE.

Could we end up feeding invalid range lengths to madvise, not page aligned? Or, could it just be inefficient if many threads try to madvise the same overlapping areas (see len calculation in os::pd_pretouch_memory)

Reply:
Small memory range and lots of workers -- that's an interesting question. I think we want to cap the chunk size from below, at least by `os::page_size()`. Let me see if I can do this without messing things up.

`os::pretouch_memory` already does `MADV_POPULATE_WRITE` when supported (see JDK-8315923). The key thing for pretouch is to eat the memory faults in multiple threads. It is arguably a kernel "issue" that `MADV_POPULATE_WRITE` is single-threaded, given that kernel can _probably_ do this with kernel workers liek we do it here on JVM side, but that is not something I expect to be available any time soon.

Reply:
Not relevant anymore, as I yanked pretouch code from this PR.